include/seqan3/alphabet/detail/alphabet_proxy.hpp

```diff
@@ -0,0 +1,292 @@
+ *
+ * ### Implementation notes
+ *
+ * The derived type needs to provide an `.impl()` member function that performs the changes in the underlying
+ * data structure.
+ *
+ * See seqan3::bitcompressed_vector or seqan3::cartesian_composition for examples of how this class is used.
+ */
+template <typename derived_type, typename alphabet_type>
+class alphabet_proxy : public alphabet_base<derived_type,
+                                            alphabet_size_v<alphabet_type>,
+                                            detail::transformation_trait_or_t<underlying_char<alphabet_type>, void>>
+{
+private:
+    //!\brief Type of the base class.
+    using base_t =  alphabet_base<derived_type,
+                                  alphabet_size_v<alphabet_type>,
+                                  detail::transformation_trait_or_t<underlying_char<alphabet_type>, void>>;
+
+    //!\brief Befriend the base type.
+    friend base_t;
+
+public:
+    // Import from base:
+    using base_t::value_size;
+    using base_t::to_rank;
+
+    /*!\name Member types
+     * \{
+     */
+    using rank_type  = underlying_rank_t<alphabet_type>;
+    using char_type  = detail::transformation_trait_or_t<underlying_char<alphabet_type>, void>;
+    using phred_type = detail::transformation_trait_or_t<underlying_phred<alphabet_type>, void>;
+    //!\}
+
+private:
+    //!\brief Never used, but required for valid definitions.
+    using char_type_virtual  = std::conditional_t<std::Same<char_type, void>, char, char_type>;
+    //!\brief Never used, but required for valid definitions.
+    using phred_type_virtual = std::conditional_t<std::Same<phred_type, void>, int8_t, phred_type>;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alphabet_proxy() : base_t{} {}
+    constexpr alphabet_proxy(alphabet_proxy const &) = default;
+    constexpr alphabet_proxy(alphabet_proxy &&) = default;
+    constexpr alphabet_proxy & operator=(alphabet_proxy const &) = default;
+    constexpr alphabet_proxy & operator=(alphabet_proxy &&) = default;
+    ~alphabet_proxy() = default;
+
+    //!\brief Construction from the emulated type.
+    constexpr alphabet_proxy(alphabet_type const a)
+    {
+        base_t::assign_rank(seqan3::to_rank(a));
+    }
+
+    //!\brief Assigment from the emulated type. This function triggers the specialisation in the derived_type.
+    constexpr derived_type & operator=(alphabet_type const & c) noexcept
+    {
+        base_t::assign_rank(seqan3::to_rank(c));
+        static_cast<derived_type &>(*this).impl(); // <- this invokes the actual proxy behaviour!
+        return static_cast<derived_type &>(*this);
+    }
+
+    //!\brief Assignment from any type that the emulated type is assignable from.
+    template <typename indirect_assignable_type>
+    constexpr derived_type & operator=(indirect_assignable_type const & c) noexcept
+        requires weakly_assignable_concept<alphabet_type, indirect_assignable_type>
```

Comment:
why is this requires after the function signature and does this not need to be guarded by `//!\cond`?

Reply:
Most of the times doxygen copes with trailing requires statements well so they don't need to be escaped. Only more complex stuff needs escaping in both cases. We realised this early on and did this in most other places, too, e.g.:
https://github.com/seqan/seqan3/blob/master/include/seqan3/alphabet/detail/member_exposure.hpp

Reply:
Thank you, good to know!